dockstore-webservice/src/main/java/io/dockstore/webservice/resources/WorkflowResource.java

```diff
@@ -1690,6 +1670,112 @@ public Workflow getWorkflowByAlias(@ApiParam(hidden = true) @Auth Optional<User>
+    public BioWorkflow addWorkflow(@ApiParam(hidden = true) @Auth User authUser,
+                                   @ApiParam(value = "Git registry", required = true, allowableValues = "GITHUB_COM, GITLAB_COM, BITBUCKET_ORG") @QueryParam("gitRegistry") TokenType gitRegistry,
+                                   @ApiParam(value = "Git repository path", required = true) @QueryParam("repository") String repository) {
+        User foundUser = userDAO.findById(authUser.getId());
+
+        // Find matching source control
+        List<Token> scTokens = checkOnBitbucketToken(foundUser)
+                .stream()
+                .filter(token -> Objects.equals(token.getTokenSource(), gitRegistry))
+                .collect(Collectors.toList());
+
+        // Add repository as workflow
+        if (scTokens.size() == 0) {
+            String msg = "User does not have access to the given source control registry.";
+            LOG.error(msg);
+            throw new CustomWebApplicationException(msg, HttpStatus.SC_BAD_REQUEST);
+        }
+
+        final Token gitToken = scTokens.get(0);
+        SourceCodeRepoInterface sourceCodeRepo = SourceCodeRepoFactory.createSourceCodeRepo(gitToken, client);
+        final String tokenSource = gitToken.getTokenSource().toString();
+
+        String gitUrl = "git@" + tokenSource + ":" + repository + ".git";
+        LOG.info("Adding " + gitUrl);
+
+        // Create a workflow
+        final Workflow createdWorkflow = sourceCodeRepo.createStubBioworkflow(repository);
+        if (createdWorkflow == null) {
+            String msg = "There was an error adding workflow with the path " + tokenSource + "/" + repository;
+            LOG.error(msg);
+            throw new CustomWebApplicationException(msg, HttpStatus.SC_BAD_REQUEST);
+        }
+
+        return saveNewWorkflow(createdWorkflow, foundUser);
+    }
+
+    /**
+     * Saves a new workflow to the database
+     * @param workflow
+     * @param user
+     * @return New workflow
+     */
+    private BioWorkflow saveNewWorkflow(Workflow workflow, User user) {
+        // Check for duplicate
+        Optional<BioWorkflow> duplicate = workflowDAO.findByPath(workflow.getWorkflowPath(), false, BioWorkflow.class);
+        if (duplicate.isPresent()) {
+            throw new CustomWebApplicationException("A workflow with the same path and name already exists.", HttpStatus.SC_BAD_REQUEST);
+        }
+        final long workflowID = workflowDAO.create(workflow);
+        final Workflow workflowFromDB = workflowDAO.findById(workflowID);
+        workflowFromDB.getUsers().add(user);
+        return (BioWorkflow)workflowFromDB;
+    }
+
+    @DELETE
+    @Timed
+    @UnitOfWork
+    @Path("/workflows")
+    @Operation(operationId = "deleteWorkflow")
+    @ApiOperation(value = "Delete a workflow for a registry and repository path.", authorizations = { @Authorization(value = JWT_SECURITY_DEFINITION_NAME) })
```

Comment:
Description should make it clear that this only works for stubs  (or that we intend it only for stubs, does restub may this grey?)